src/expr/src/explain/text.rs

```diff
@@ -298,15 +298,52 @@ impl MirRelationExpr {
                     })?;
                 }
             }
-            LetRec { ids, values, body } => {
-                let bindings = ids.iter().zip(values.iter()).collect::<Vec<_>>();
+            LetRec {
+                ids,
+                values,
+                max_iters,
+                body,
+            } => {
+                assert_eq!(ids.len(), values.len());
+                assert_eq!(ids.len(), max_iters.len());
+                let bindings = itertools::izip!(ids.iter(), values.iter(), max_iters.iter())
+                    .collect::<Vec<(&LocalId, &MirRelationExpr, &Option<u64>)>>(); // CLion needs these explicit types
                 let head = body.as_ref();
 
+                // Determine whether all `max_iters` are the same.
+                // If all of them are the same, then we'll print it on top of the block (or not
+                // print it at all of it's None). If there are differences, then we print them on
+                // the ctes.
+                let all_max_iters_same = {
+                    let mut all_same = true;
+                    let first = max_iters.iter().next().unwrap();
+                    for max_iter in max_iters {
+                        if max_iter != first {
+                            all_same = false;
+                        }
+                    }
+                    if all_same {
+                        first
+                    } else {
+                        &None
+                    }
+                };
```

Comment:
This looks like it can be a single `reduce`:
```rust
let all_max_iters_same = max_iter.iter().reduce(|first, i| {
    if i == first { first } else { &None }
});
```

Reply:
Thanks!